impl/maven-impl/src/main/java/org/apache/maven/impl/DefaultModelXmlFactory.java

```diff
@@ -48,8 +48,37 @@
 @Singleton
 public class DefaultModelXmlFactory implements ModelXmlFactory {
     @Override
+    @Nonnull
     public Model read(@Nonnull XmlReaderRequest request) throws XmlReaderException {
         nonNull(request, "request");
+        Model model = doRead(request);
+        if (isModelVersionGreaterThan400(model)
+                && !model.getNamespaceUri().startsWith("http://maven.apache.org/POM/")) {
+            throw new XmlReaderException(
+                    "Invalid namespace '" + model.getNamespaceUri() + "' for model version " + model.getModelVersion(),
+                    null,
+                    null);
+        }
+        return model;
+    }
+
+    private boolean isModelVersionGreaterThan400(Model model) {
+        String version = model.getModelVersion();
+        if (version == null) {
+            return false;
+        }
+        try {
```

Comment:
There's no need to do all this parsing. The namespace URI should be a fixed, known string and that remains true in future versions. We should not change the namespace URI in new versions.

Reply:
That's not the point. The point is that the reader needs to support existing models which have no namespace, and that cannot be enforced because they are in Maven Central.  So we can only enforce for model > 4.0.0.

Reply:
My mistake. I was confusing model version with the namespace URI